Establish a new error handling project group (#2965)
rust-lang/rfcs: text/0000-project-error-handling.md

- Feature Name: error_handling_project_group
- Start Date: 2020-07-23
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

This RFC establishes a new project group, under the libs team, to drive efforts to improve error handling in Rust.

# Motivation
[motivation]: #motivation

> The error handling project group aims to reduce confusion on how to structure error handling for users in the rust community. This will be accomplished by creating learning resources and pushing effort to upstream widely used crates into std. As a secondary goal this project group will also try to resolve some known issues with the Error trait and reporting errors in panics/termination.

Reviewer: ```suggestion
The error handling project group aims to reduce confusion on how to structure error handling for users in the Rust community. This will be accomplished by creating learning resources and pushing effort to upstream widely used crates into the standard library. As a secondary goal, this project group will also try to resolve some known issues with the `Error` trait and reporting errors in panics/termination.
```

# Charter
[charter]: #charter

## Goals

### Agree on and define common error handling terminology

- Recoverable error: An error that can reasonably be expected to be encountered e.g. a missing file.
- Unrecoverable error: An error that cannot reasonably be expected to happen and which indicates a bug e.g. indexing out of bounds.

> - Error Type: A type that represents a recoverable error. Error types can optionally implement the error trait so that it can be reported to the user or be converted into a trait object.

Reviewer: ```suggestion
- Error Type: A type that represents a recoverable error. Error types can optionally implement the `Error` trait so that it can be reported to the user or be converted into a trait object.
```

> - Reporting Type: A type that can store all recoverable errors an application may need to propogate and print them as error reports.

Reviewer: ```suggestion
- Reporting Type: A type that can store all recoverable errors an application may need to propagate and print them as error reports.
```

I suggest running a spell checker locally to catch other cases that people didn't find.

>     - Reporting types can represent the recoverable errors either via concrete types, likely paramaterized, or trait objects.

Reviewer: ```suggestion
    - Reporting types can represent the recoverable errors either via concrete types, likely parameterized, or trait objects.
```

    - Reporting types often bundle context with errors when they are constructed, e.g. `Backtrace`.
    - Reporting types often provide helper functions for creating ad hoc errors whose only purpose is to be reported e.g. `anyhow::format_err!` or `eyre::WrapErr`.

### Come to a consensus on current best practices

Here is a tenative starting point, subject to change:

- Use `Result` and `Error` types for recoverable errors.

> - Use `panic` for unrecoverable errors.

Reviewer 2: Do you want to touch `unwind`? I am asking, because that is for `no_std` and `async` [poorly](https://doc.rust-lang.org/nomicon/unwinding.html)[documented](https://rust-embedded.github.io/book/start/panicking.html)
and let to some confusion [here](https://github.com/rust-lang/rfcs/pull/2958).

Author: I'm not particularly well informed on the implementation details of unwinding but I can see it being within the purview of this project group.

> - Impl `Error` for error types that may need to be reported to a human or be composed with other errors.

Reviewer: ```suggestion
- Implement `Error` for error types that may need to be reported to a human or be composed with other errors.
```

- Use enums for types representing multiple failure cases that may need to be handled.

>     - For libraries, oftentimes you want to support both reporting and handling so you impl `Error` on a possibly non-exhaustive enum.

Reviewer: ```suggestion
    - For libraries, oftentimes you want to support both reporting and handling so you implement `Error` on a possibly non-exhaustive enum.
```

- Error kind pattern for associating context with every enum variant without including the member in every enum variant.
- Convert to a reporting type when the error is no longer expected to be handled beyond reporting e.g. `anyhow::Error` or `eyre::Report` or when trait object + downcast error handling is preferable.
- Recommend `Box`ing concrete error types when stack size is an issue rather than `Box`ing and converting to `dyn Error`s.

> - What is the concensus on handling `dyn Error`s? Should it be encouraged or discouraged? Should we look into making `Box<dyn Error...>` impl Error?

Reviewer 3: s/concensus/consensus/

> - What is the concensus on handling `dyn Error`s? Should it be encouraged or discouraged? Should we look into making `Box<dyn Error...>` impl Error?

Author: ```suggestion
- What is the consensus on handling `dyn Error`s? Should it be encouraged or discouraged? Should we look into making `Box<dyn Error...>` impl Error?
```

> - What is the consensus on handling `dyn Error`s? Should it be encouraged or discouraged? Should we look into making `Box<dyn Error...>` impl Error?

Reviewer: ```suggestion
- What is the consensus on handling `dyn Error`s? Should it be encouraged or discouraged? Should we look into making `Box<dyn Error...>` implement `Error`?
```

Reviewer: This isn't possible to do the 2 or 3 times I've attempted it.

### Communicate current best practices

> - Document the concensus.

Reviewer: ```suggestion
- Document the consensus.
```

- Communicate plan for future changes to error handling, and the libraries that future changes are being based off of.
- Produce learning resources related to current best practices.
    - New chapters in the book?

> ### Evaluate options for error reporting type aka better `Box<dyn Error>`

Reviewer: ```suggestion
### Evaluate options for error reporting type a.k.a. better `Box<dyn Error>`
```

- Survey the current libraries in the ecosystem:
    - `anyhow`
    - `eyre`
- Evaluate value of features including:
    - Single word width on stack
    - Error wrapping with display types and with special downcast support.

>     - report hook + configurable `dyn ReportHandler` type for custom report formats and content, similar to panic handler but for errors.

Reviewer: ```suggestion
    - Report hook and configurable `dyn ReportHandler` type for custom report formats and content, similar to panic handler but for errors.
```

>     - libcore compatibility.

Reviewer: Everything should be compatible with libcore. I think you want to phrase this as "doesn't require libstd"

Author: how about

```suggestion
    - `#[no_std]` support
```

### Consolidate ecosystem by merging best practice crates into std

- Provide a derive macro for `Error` in std.

> - Stabilize the `Backtrace` type but possibly not `fn backtrace` on the Error trait.

Reviewer: ```suggestion
- Stabilize the `Backtrace` type but possibly not `fn backtrace` on the `Error` trait.
```

    - Provide necessary API on `Backtrace` to support crates like `color-backtrace`.

> - Move error to core .

Author: related: https://github.com/rust-lang/rust/pull/72981#issuecomment-663191897

> - Move error to core .

Reviewer: ```suggestion
- Move `Error` to core.
```

    - Depends on generic member access.
    - Requires resolving downcast dependency on `Box` and blocking the stabilization of `fn backtrace`.

> - Potentially stabilize an error Reporting type based on `anyhow` and `eyre` now that they're close to having identical feature sets.

Reviewer: ```suggestion
- Potentially stabilize an error reporting type based on `anyhow` and `eyre` now that they're close to having identical feature sets.
```

### Add missing features

- Generic member access on the `Error` trait.
- `Error` return traces:
    - Depends on specialization and generic member access.
- Fix rough corners around reporting errors and `Termination`.

## Non Goals

> - This group should not be involved in design discussions for the `Try` trait, `try` blocks, or `try` fns.

Reviewer: "Involved" doesn't appear to be the right word. The way this is phrased, it sounds as if **members** of this group are not allowed to participate in the discussion.

Author: ```suggestion
- This group should not be managing in design discussions for the `Try` trait, `try` blocks, or `try` fns.
```

Hows this?

Reviewer 4: Another option: add a parenthetical to the effect of "(Not that people can't participate in both this group and in `try` stuff)"

## Membership Requirements

> - Group membership is open, any interested party can participate in discussions, repeat contributors will be added appropriate teams.

Reviewer: ```suggestion
- Group membership is open, any interested party can participate in discussions, repeat contributors will be added to appropriate teams.
```

## Additional Questions

### What support do you need, and separately want, from the Rust organization?

I'm not sure, my main concern is getting prompt feedback on RFCs.

### Why should this be a project group over a community effort?

> Error handling has historically been handled via community effort which has resulted a great deal of competing and outdated solutions and resources. An officially supported project group has the best chance of writing and maintaining up-to-date resources that would reach a wide audience.

Reviewer: I don't know that I fully agree with this declaration. 

- "a great deal of competing [...] solutions" is presented as a negative, but I don't see any details in this proposal to support the claim.
- "outdated [...] resources" — I'm not the target of these resources so I'm not familiar with any outdated resources. Could you provide some links?
- "An officially supported project group has the best chance of writing and maintaining up-to-date resources" — What prevents creating these resources, keeping them updated for "long enough", and then having them be adopted?

Since you said "solutions and resources", I'm interpreting it as solution == "a crate" and resource == "documentation". If that's not what you mean, you may wish to clarify.

Author: You raise good points, I'll try to re-evaluate the justification here. By outdated resources I was recalling https://blog.burntsushi.net/rust-error-handling/ but it's actually been updated this year, you're right that this seems to mostly be FUD.

- Feature Name: error_handling_project_group
- Start Date: 2020-07-23
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

This RFC establishes a new project group, under the libs team, to drive efforts to improve error handling in Rust.

# Motivation
[motivation]: #motivation

The error handling project group aims to reduce confusion on how to structure error handling for users in the Rust community. This will be accomplished by creating learning resources and pushing effort to upstream widely used crates into the standard library. As a secondary goal, this project group will also try to resolve some known issues with the `Error` trait and reporting errors in panics/termination.

# Charter
[charter]: #charter

## Goals

### Agree on and define common error handling terminology

- Recoverable error: An error that can reasonably be expected to be encountered e.g. a missing file.

> - Unrecoverable error: An error that cannot reasonably be expected to happen and which indicates a bug e.g. indexing out of bounds.

Reviewer 5: Hm, why define "recoverable" and "unrecoverable" errors by how *likely* they're thought to be rather than by whether it's feasible for software to recover from them?

In particular, the subsequent recommendation to use "`Result` and `Error` types for recoverable errors" and "`panic` for unrecoverable errors" would seem, at least to me, to agree better with the latter, more literal interpretation of the terms "recoverable" and "unrecoverable".

- What is the consensus on handling `dyn Error`s? Should it be encouraged or discouraged? Should we look into making `Box<dyn Error...>` implement `Error`?

### Identify pain points in error handling today

- Backtrace capture is expensive, but without it can be difficult to pinpoint the origin of errors
- unwrapping errors without first converting to a reporting type will often discard relevant information
- errors printed from `main` have to assume a prefixed `Error: `, sub-par control of output format when printing during termination.
- The `Error` trait only exposes three forms of context, can only represent singly-linked lists for chains of errors

### Communicate current best practices

- Document the consensus.
- Communicate plan for future changes to error handling, and the libraries that future changes are being based off of.
- Produce learning resources related to current best practices.
    - New chapters in the book?

### Evaluate options for error reporting type a.k.a. better `Box<dyn Error>`

- Survey the current libraries in the ecosystem:
    - `anyhow`

>     - `eyre`

Reviewer 5: Does `snafu` belong in this list?

Author: Not as an error reporter, snafu is used to define error types that implement the error trait, not to define logic that uses the error trait to print the chain of errors and associated context like the backtrace.

### What support do you need, and separately want, from the Rust organization?

I'm not sure, my main concern is getting prompt feedback on RFCs.

### Why should this be a project group over a community effort?

There isn't anything in this project group that can't be handled as a
community effort, but centralizing work into a project group should help
speed things. Error handling is a core aspect of the language and changes in
error handling have large impacts on the ecosystem. Ensuring that efforts to
refine error handling within Rust have sufficient resources and don't stall
out is in the best interests of the community. By organizing efforts as a
project group we will hopefully have an easier time recruiting new members,
getting attention on RFCs from members of the libs team, and using the
established resources and expertise of the rust organization for coordinating
our efforts.

### What do you expect the relationship to the team be?

The project group will create RFCs for various changes to the standard library and the team will review them via the standard RFC process.

### Who are the initial shepherds/leaders? (This is preferably 2–3 individuals, but not required.)

> Jane Lusby, Andrew Gallant, and Sean Chen.

Reviewer 5: I'd like to suggest including the relevant GitHub usernames here.  (I'm afraid I don't know who Sean Chen is.)

- What is the consensus on handling `dyn Error`s? Should it be encouraged or discouraged? Should we look into making `Box<dyn Error...>` implement `Error`?

### Identify pain points in error handling today

> - Backtrace capture is expensive, but without one it can be difficult to pinpoint the origin of errors

Reviewer: ```suggestion
- Backtrace capture is expensive, but without it can be difficult to pinpoint the origin of errors
```

Reviewer: I generally disagree with the universal applicability of this point as well; this is a key point of SNAFU — producing a semantic backtrace via the error types.

Author: Assuming I understand what you mean, I think that the semantic backtraces via `SNAFU` can definitely help here but I don't think they're a universal solution. It still fairly easy to misuse. For example, if a developer introduces a higher level snafu error to handle an `io::Error`, but then uses this type liberally throughout the code. Unless you carefully create one variant per source location the origin of errors can still be lost.

Here's an example of this kind of thing going wrong via `thiserror`. https://github.com/ZcashFoundation/zebra/issues/758

> - unwrap on errors without first converting to a reporting type will often discard relevant information

Reviewer: ```suggestion
- unwrapping errors without first converting to a reporting type will often discard relevant information
```

> - errors printing from main have to assume a prefixed `Error: `, sub par control of output format when printing during termination.

Reviewer: ```suggestion
- errors printed from `main` have to assume a prefixed `Error: `, sub-par control of output format when printing during termination.
```

Reviewer: I don't know what "have to assume a prefixed `Error: `" means.

Author: I'm referring to this default in `std`'s impl of `Termination` for `Result` https://doc.rust-lang.org/stable/src/std/process.rs.html#1690-1696, right now there's no way to signal to `Debug` types that they're being formatted as part of `Termination` or not. @dtolnay has suggested adding an `is_termination` flag to `Debug`. I'm also interested in the possibility of a 3rd `std::fmt` trait called `std::fmt::Report` to handle this, leveraging specialization.

> - Error trait only exposes 3 forms of context, can only represent singly linked lists for chains of errors

Reviewer: ```suggestion
- The `Error` trait only exposes three forms of context, can only represent singly-linked lists for chains of errors
```

Reviewer: It's unclear what the pain points are here. 

- Is three too many or not enough?
- Why is a singly-linked list painful?

Author: 3 is not enough, I'd like to be able to return `SpanTraces`, `std::panic::Location`, and other such types via the error trait, this is a reference to the generic member access RFC.

Singly linked lists can be painful for parsers for example, where you encounter multiple errors at once, and they're all the source for a single higher level error, so your chain of errors looks a little more like a tree. This is also addressed by the generic member access proposal where you could access a `dyn Iter<&(dyn Error + 'static)>`.
